api/src/main/java/org/apache/iceberg/actions/RewriteTablePath.java

```diff
@@ -0,0 +1,91 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.actions;
+
+import org.apache.iceberg.Table;
+
+public interface RewriteTablePath extends Action<RewriteTablePath, RewriteTablePath.Result> {
+
+  /**
+   * Passes the source and target prefixes that will be used to replace the source prefix with the
+   * target one.
+   *
+   * @param sourcePrefix the source prefix to be replaced
+   * @param targetPrefix the target prefix
+   * @return this for method chaining
+   */
+  RewriteTablePath rewriteLocationPrefix(String sourcePrefix, String targetPrefix);
+
+  /**
+   * Pass the version copied last time. It is optional if the target table is provided. The default
+   * value is the target table's current version. User needs to make sure whether the start version
+   * is valid if target table is not provided.
+   *
+   * @param lastCopiedVersion only version file name is needed, not the metadata json file path. For
+   *     example, the version file would be "v2.metadata.json" for a Hadoop table. For metastore
+   *     tables, the version file would be like
+   *     "00001-8893aa9e-f92e-4443-80e7-cfa42238a654.metadata.json".
+   * @return this for method chaining
+   */
+  RewriteTablePath lastCopiedVersion(String lastCopiedVersion);
+
+  /**
+   * The latest version of the table to copy. It is optional, the default value is the source
+   * table's current version.
+   *
+   * @param endVersion only version file name is needed, not the metadata json file path. For
+   *     example, the version file would be "v2.metadata.json" for a Hadoop table. For metastore
+   *     tables, the version file would be like
+   *     "00001-8893aa9e-f92e-4443-80e7-cfa42238a654.metadata.json".
+   * @return this for method chaining
+   */
+  RewriteTablePath endVersion(String endVersion);
+
+  /**
+   * Set the customized staging location. It is optional. By default, staging location is a
+   * subdirectory under table's metadata directory.
+   *
+   * @param stagingLocation the staging location
+   * @return this for method chaining
+   */
+  RewriteTablePath stagingLocation(String stagingLocation);
+
+  /**
+   * Set the target table. It is optional if the start version is provided.
+   *
+   * @param targetTable the target table
+   * @return this for method chaining
+   */
+  RewriteTablePath targetTable(Table targetTable);
+
+  /** The action result that contains a summary of the execution. */
+  interface Result {
```

Comment:
Actually a few more comments here, I think we need additional list for 

- position delete files
- equality delete files
- puffin files
Does it make sense?

Also, what do you guys think to have a source, target path pair.  Each source path is either the original file in the table, or a rewritten one in staging.  The target path is the one under target prefix (the ones that are now referred to by all rewritten metadata paths).  I know in the earlier implementation this existed.

cc @flyrain for thoughts

Reply:
@szehon-ho the I don't think it makes sense, because `metadataFileList` file will contain all metadata files paths (metadata.json, manifests, manifest lists, position/equality delete files etc), so there's no point of having multiple lists. I agree with returning source and target path for each file, will keep that in mind for the implementation PR.

Reply:
yea i thought a bit over the weekend, and multiple list is too complicated.

How about we dont need multiple list (metadataFileList and dataFileList), and just have one list (fileList).  Because its ambiguious if deleteFile , which list it should be part of.  In the end, we want to have just a list of Pair<String, String> , which is source and target.  Source can be source if rewritten, or original source if not.

Reply:
Since `deleteFiles` already exists under the `metadata` folder in the storage then it makes sense to have them in the `metadataFileList`. I think the expectation was for the user to copy all files in `metadataFileList ` to `/metadata` and `dataFileList` to `/data` in the new target location. But I agree with your suggestion, we can further narrow it down to one list as well and just have the source,target paths for all of them and the user can just initiate the copy from source to target without worrying about if it's data or metadata files 👍

Reply:
> Since deleteFiles already exists under the metadata folder in the storage

I thought its generally not the case, and they co-exist with the data files in the partition folder.

Reply:
DeleteFiles are not considered as metadata files even they are in the directory `/metadata`. It's more and more clear to me that we may just use one list holding pairs of (sourcePath, targetPath). WDYT?